### PR TITLE
Use different name table offset depending on staff type

### LINKED ISSUE #22809: Reused "real" names for staff
### Operating System

Windows 11, 64-bit

### OpenRCT2 build

OpenRCT2, v0.4.14-59-g4b44f4f (4b44f4f on develop) provided by GitHub

### Base game

RollerCoaster Tycoon 2

### Area(s) with this issue?

This is a development issue

### Describe the issue

As requested by @AaronVanGeffen for issue tracking:

The issue concerns the "show real name" options. When enabled:

1) The same names are reused for different types of staff. For instance, the first handyman will be named Bernard C. and the first mechanic is also named the same. This also applies to the security guard and entertainer staff. Subsequently, the next name, Lindsey C. is used for all the second staff of each type.

2) The same names are always used. First staff is always named Bernard, second is Lindsey, third is Cynthia, and etc. I tested with different scenarios and the names used are in that same ordered. Would it be possible to assign their names more randomly?

### Steps to reproduce

1. Enable show real name of staff in options
2. Hire any staff type
3. Hire a different staff type
4. Observe their assigned names and notice they are the same

### Attachments

_No response_

### PR DESCRIPTION
**Pull request by AaronVanGeffen**
Should fix #22809.

### CODE REVIEW

**Comment by fidwell** [comment]
I gave this a quick test, and it is better, but still duplicates names such that handymen have all names, and other staff are duplicates of those handymen.

The first four handymen are:
1. Bernard C.
2. Cynthia C. — also Security Guard 1
3. Lindsey C. — also Mechanic 1
4. Nellie C. — also Entertainer 1

**Comment by AaronVanGeffen** [comment]
> I gave this a quick test, and it is better, but it still assigns names such that handymen have all names, and other staff are duplicates of those handymen.

Ahh, I didn't realise it was only offset this little. I've increased the offset by an additional factor of 256. Is that working better?

> Also (and this is just personal preference) it's a bit weird to have the staff names _always_ start with Bernard C. [...] It might be nice to randomize that, or at least make it appear more random. Though (if I'm not mistaken) since the guest names are generated programmatically this may not be possible yet.

You're right, it would be better to randomize the offset. I'm not certain what the best approach would be, to be honest. Perhaps generating a one-time random offset when a scenario is started? That could apply to both guests and staff.

**Comment by fidwell** [comment]
Yes, this is a lot better. If the names only repeat after 256 staff are hired, this should be more than sufficient.

image

> Perhaps generating a one-time random offset when a scenario is started? That could apply to both guests and staff.

Yeah that seems like a promising idea. But I'd say it's out-of-scope for this PR, so I'd call this change good, as it is currently.

**Review by fidwell — APPROVED** [review]

File: src/openrct2/entity/Peep.cpp [hunk 1]
```diff
@@ -1608,7 +1608,11 @@ void Peep::FormatNameTo(Formatter& ft) const
 
         if ((!isStaff && showGuestNames) || (isStaff && showStaffNames))
         {
-            auto realNameStringId = GetRealNameStringIDFromPeepID(PeepId);
+            auto nameId = PeepId;
+            if (isStaff)
+                nameId *= 256 * EnumValue(staff->AssignedStaffType) + 1;
+
+            auto realNameStringId = GetRealNameStringIDFromPeepID(nameId);
```
Comment (tupaschoal) [inline src/openrct2/entity/Peep.cpp:1615]: Is the ternary much here? :D
```suggestion
            auto nameId = isStaff ? (PeepId * 256 * EnumValue(staff->AssignedStaffType) + 1) : PeepId;
            auto realNameStringId = GetRealNameStringIDFromPeepID(nameId);
```

In any case, I think a comment stating what this offset is doing will be good for future readers.

**Review by tupaschoal — APPROVED** [review]

File: src/openrct2/entity/Peep.cpp (hunk 1 above)
Comment (AaronVanGeffen) [reply src/openrct2/entity/Peep.cpp:1615]: Haha, I considered a ternary, but I don't think it improves the readability tbh.

Agreed a comment would be good. I'll add one tomorrow.

File: src/openrct2/entity/Peep.cpp [hunk 2]
```diff
@@ -1608,7 +1608,11 @@ void Peep::FormatNameTo(Formatter& ft) const
 
         if ((!isStaff && showGuestNames) || (isStaff && showStaffNames))
         {
-            auto realNameStringId = GetRealNameStringIDFromPeepID(PeepId);
+            auto nameId = PeepId;
+            if (isStaff)
+                nameId *= 256 * EnumValue(staff->AssignedStaffType) + 1;
```
Comment (ZehMatt) [inline src/openrct2/entity/Peep.cpp:1613]: Maybe its better to use a prime to multiply so there is less of a pattern.

Comment (AaronVanGeffen) [reply src/openrct2/entity/Peep.cpp:1613]: What kind of prime are you thinking? Close primes are 251, 257, 263, but I don't think that would make too much of a difference with how this is currently set up.

Comment (fidwell) [reply src/openrct2/entity/Peep.cpp:1613]: I'm not sure it makes a difference, either. If [this](https://youtu.be/12EqP3VL8FY?t=79) is still accurate and there are 16,384 total names, then to get the most spread over available names for staff, you'd want to offset each staff type by 4,096.

File: src/openrct2/entity/Peep.cpp (hunk 1 above)
Comment (AaronVanGeffen) [reply src/openrct2/entity/Peep.cpp:1615]: @tupaschoal Does this sound good to you?
```suggestion
            auto nameId = PeepId;
            if (isStaff)
            {
                // Prevent staff from getting the same names by offsetting the name table based on staff type.
                nameId *= 256 * EnumValue(staff->AssignedStaffType) + 1;
            }

            auto realNameStringId = GetRealNameStringIDFromPeepID(nameId);
```

Comment (tupaschoal) [reply src/openrct2/entity/Peep.cpp:1615]: It does!

**Pull request merged**